baselines/fedpac/fedpac/client.py

```diff
@@ -0,0 +1,240 @@
+        num_epochs: int,
+        learning_rate: float,
+        lamda: float,
+
+    ):
+        self.net = net
+        self.trainloader = trainloader
+        self.valloader = valloader
+        self.device = device
+        self.num_epochs = num_epochs
+        self.learning_rate = learning_rate
+        self.lamda = lamda
+        self.num_classes = self.net.num_classes
+        self.feature_extractor = self.get_feature_extractor()
+        self.feature_centroid = get_centroid(self.feature_extractor)
+        self.class_sizes = self.get_class_sizes()
+        self.class_fractions = self.get_class_fractions()
+
+    def get_parameters(self, config: Dict[str, Scalar]) -> NDArrays:
+        """Returns the parameters of the current net."""
+        return [val.cpu().numpy() for _, val in self.net.state_dict().items()]
+
+    def set_parameters(self, parameters: NDArrays) -> None:
+        """Changes the parameters of the model using the given ones."""
+        params_dict = zip(self.net.state_dict().keys(), parameters)
+        state_dict = OrderedDict({k: torch.Tensor(v) for k, v in params_dict})
+        self.net.load_state_dict(state_dict, strict=True)
+
+
+    def get_class_sizes(self):
+        dataloader = self.trainloader
+        sizes = torch.zeros(self.num_classes)
+        for images, labels  in dataloader:
+            for i in range(self.num_classes):
+                sizes[i] = sizes[i] + (i == labels).sum()
+        return sizes
+
+    def get_class_fractions(self):
+        total = len(self.trainloader)
+        return self.class_sizes / total
+
+
+    def get_statistics(self):
+        dim = self.net.state_dict()[self.net.classifier_layers[0]][0].shape[0] 
+        feat_dict = self.feature_extractor
+        for k in feat_dict.keys():
+            feat_dict[k] = torch.stack(feat_dict[k])
+        
+        py = self.get_class_fractions()
+        py2 = torch.square(py)
+        v = 0
+        h_ref = torch.zeros((self.num_classes, dim), device=self.device)
+        datasize = torch.tensor(len(self.trainloader)).to(self.device)
+        for k in feat_dict.keys():
+            feat_k = feat_dict[k]
+            num_k = feat_k.shape[0]
+            feat_k_mu = feat_k.mean(dim=0)
+            h_ref[k] = py[k]*feat_k_mu
+            v += (py[k]*torch.trace((torch.mm(torch.t(feat_k), feat_k)/num_k))).item()
+            v -= (py2[k]*(torch.mul(feat_k_mu, feat_k_mu))).sum().item()
+        v = v/datasize.item()
+        return (v, h_ref)
+
+        
+    
+    def get_feature_extractor(self):
+        """function to extract feature extractor layers"""
+        feature_extractors = {}
+        model=self.net
+        train_data= self.trainloader
+        with torch.no_grad():
+            for inputs, labels in train_data:
+                inputs, labels = inputs.to(self.device), labels.to(self.device)
+                features, outputs = model(inputs)
+                feature_extractor = features.clone().detach()
```

Comment:
is a `.cpu()` missing here? I'm getting an error in the strategy when clients train on GPU. (which seem to return centroids that are not on `CPU`).